examples/all-clusters-app/all-clusters-common/gen/attribute-size.cpp

```diff
@@ -31,15 +31,36 @@ using namespace chip::app::List;
 // and data is undefined.
 constexpr uint16_t kSizeLengthInBytes = 2u;
 
+#if BIGENDIAN_CPU
+uint8_t * reverseMemcpy(uint8_t * dest, const uint8_t * src, size_t len)
+{
+    uint8_t * d       = dest + len - 1;
+    const uint8_t * s = src;
+    while (len--)
+    {
+        *d-- = *s++;
+    }
+    return dest;
+}
+#endif
+
 void copyListMember(uint8_t * dest, uint8_t * src, bool write, uint16_t * offset, uint16_t length)
 {
     if (write)
     {
+#if BIGENDIAN_CPU
+        reverseMemcpy(dest + *offset, src, length);
```

Comment:
Several things:

* We shouldn't replace a memmove with a memcpy unless we know for sure this is safe.  And if it is safe, then all of the memmoves should be changed to memcpy.
* This is being introduced in a generic way that implies all zcl arrays enclose arithmetic types that have an endianness in the first place.  Is this true?
* If this is mostly to manage a uint16 length, why aren't we just using `ntoh`?

Reply:
Agree that we should consider changing memmove's to memcpy's, since all of these are working on two non-intersecting memory buffers. @vivien-apple can you tell us why memmove's were used here in your original commit?

We could skip this reverse-copy method for single-byte entities (although it probably wouldn't save much is execution speed). The copyList implementation which is the exclusive caller of this function uses this for data types with the exception of strings, in which case it uses ByteSpan.

The uint16 length management does not use this function. (see lines 64..69 of this file)

Reply:
> We shouldn't replace a memmove with a memcpy unless we know for sure this is safe

Nothing _really_ guarantees src and dest are non-overlapping, though in practice they are, I think.

> Is this true?

It's ... complicated.  We support arrays of structs, but we handle those by calling copyListMember on each struct member individually.  That said, if a struct member is itself a struct, or an array, I suspect bad things would happen.  I don't know whether we support those cases yet.

Reply:
So we have two considerations here: (1) Do we want to support callers that pass overlapping buffers for in-place manipulation?  (2) Should behavior at least be consistent?

I argue that (2) is obvious.  If any of the paths use memmove, they all should support overlapping buffers and the revereseMemcpy needs to be reworked into a reverseMemmove.  Else we should change everything to memcpy.

I defer to everybody else on whether we need memmove or memcpy, but there is no reasonable defense for mixing the two.   What we currently have in this PR (mempcy for a handful of operations on big-endian systems, memmove everywhere else) is a vehicle to inject subtle bugs into our minority big-endian platforms.

Reply:
OK, here I think is the pragmatic answer.

Consider:

`uint16_t emberAfCopyList(ClusterId clusterId, EmberAfAttributeMetadata * am, bool write, uint8_t * dest, uint8_t * src, int32_t index)`

These `emberAfCopyList`  functions currently appear to be using memmove through their call chains, including calls into `copyListMember`.  However, copies are internally achieved element-by-element.  Thus, in-place manipulation can only be correct if `src` and `dest` are the same address and `dest` output will be the same size as input, OR the buffers are totally non-overlapping.  Is that functionality we would use or need?  That is a much smaller set of supportable use cases than we get with a single memmove call, which will have correct behavior for **any** overlapping buffers.  And if we do need to support this very narrow overlapping buffer use case, that really warrants some pretty detailed comments in the code on what is legal and what is not.

More likely, it appears the use of memmove was thought to confer more safety that it actually does, or was selected to avoid having to immediately perform this type of analysis.  Given this, I don't see the utility of retaining these memmoves unless there is a significant refactor to actually support overlapping buffers.  And removing the well-intentioned, but ultimately unneeded memmoves sounds like a followup task.  Given that, this PR's approach with reverseMemcpy seems reasonable.

Thoughts?

Reply:
I think semantically having these all be memcpy, not memmove makes sense.  And I agree we should make the behavior be consistent on big- and little-endian to avoid subtle behavior differences.

Reply:
As for the issue of whether `copyListMember` is safe to use on the type passed to it, I wonder whether we could have some sort of static asserts somewhere at the callsites that (1) we know the type and (2) it's safe to just byte-reverse that type.

Reply:
OK, filed issue for memmove -> memcpy refactor here: https://github.com/project-chip/connectedhomeip/issues/7492

For the remaining issue (type correctness), do we wish to have asserts here in this PR to verify that byte-swapped fields are actually arithmetic types where this is appropriate, or do we want that as a separate issue as well?

Reply:
Figuring out how to assert that is not trivial: it needs to be done in the caller of this method, not in the method itself, right?

Reply:
Sure.  True.  reverseMemcpy can't know whether its caller is doing the right thing.  Maybe we just table it.  Is this PR calling it for the right cases? I assume it is.

Reply:
I think it is, but would love @vivien-apple to give it a once-over when he gets a chance to.